Auto select identity based on X-Original-To / Delivered-To / X-Envelope-To / TO / CC

YESALAM opened the pull request:
As @Valodim suggested here checking for X-Original-To and Delivered-To also before To and Cc which would solve Issue #1407 .

YESALAM commented:
Thanx for review .  I have added the required tests .

YESALAM commented:
Hey @philipwhiuk would you please review the changes and give  me feedback on this .

cketti commented on k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java:
```diff
@@ -48,6 +48,9 @@
     protected Address[] mCc;
     protected Address[] mBcc;
     protected Address[] mReplyTo;
+    protected Address[] mX_OriginalTo;
```
We're moving away from using the `m` prefix for fields. Please don't use it for new fields.
Also, please don't use underscores in field names. This should be

```java
protected Address[] xOriginalTo;
protected Address[] deliveredTo;
protected Address[] xEnvelopeTo;
```

cketti commented on k9mail-library/src/test/java/com/fsck/k9/mail/internet/MimeMessageParseTest.java:
```diff
@@ -272,4 +273,60 @@ private static void checkLeafParts(MimeMessage msg, String... expectedParts) thr
     }
+
+    @Test
+    public void isXOriginalTo() throws Exception {
```
This name doesn't describe what is being tested. Suggestion: `getRecipients_withXOriginalTo()`.

cketti commented on k9mail-library/src/test/java/com/fsck/k9/mail/internet/MimeMessageParseTest.java:
```diff
@@ -272,4 +273,60 @@ private static void checkLeafParts(MimeMessage msg, String... expectedParts) thr
+
+        Address address = new Address("<test@mail.com>");
+        Address[] x_original_address = msg.getRecipients(RecipientType.X_ORIGINAL_TO) ;
+        assertTrue(x_original_address[0].equals(address));
```
If this test fails the error message generated by JUnit won't be too helpful. Change to the following instead:

```java
assertEquals(new Address("<test@mail.com>"), xOriginalAddresses[0]);
```

Before that you should probably check the number of returned addresses.

```java
assertEquals(1, xOriginalAddresses.length);
```

cketti commented on k9mail-library/src/test/java/com/fsck/k9/mail/internet/MimeMessageParseTest.java:
```diff
@@ -272,4 +273,60 @@ private static void checkLeafParts(MimeMessage msg, String... expectedParts) thr
+                        "this is some test text."));
+
+        Address address = new Address("<test@mail.com>");
+        Address[] x_original_address = msg.getRecipients(RecipientType.X_ORIGINAL_TO) ;
```
Local variables shouldn't contain underscores either. Make this `xOriginalAddresses` (plural, because it's an array).

cketti commented on k9mail/src/androidTest/java/com/fsck/k9/helper/IdentityHelperTest.java:
```diff
@@ -0,0 +1,141 @@
+import java.util.List;
+
+/**
+ * Created by yesalam on 7/3/17.
```
Please remove this comment.

cketti commented on k9mail/src/androidTest/java/com/fsck/k9/helper/IdentityHelperTest.java:
```diff
@@ -0,0 +1,141 @@
+ * Created by yesalam on 7/3/17.
+ */
+
+public class IdentityHelperTest extends ApplicationTestCase<K9> {
```
I don't see why this needs to be an instrumentation test. It should run fine using Robolectric on the JVM.

cketti commented on k9mail/src/androidTest/java/com/fsck/k9/helper/IdentityHelperTest.java:
```diff
@@ -0,0 +1,141 @@
+
+
+    @Override
+    protected void setUp() throws Exception {
```
Much of this initialization is unnecessary.

YESALAM commented:
I changed tests from AndroidTest to Roboelectric which is way better than I thought .
Also I have modified the code style by importing k-9 code style.

philipwhiuk commented:
Thanks @YESALAM - I've rebased and merged this in under #2454
